crates/provider/src/helpers.rs

```diff
@@ -0,0 +1,180 @@
+use std::{error::Error, fs::File};
+
+use crate::{native::FileSystem, shared::types::LocalFile};
+
+#[derive(Debug, PartialEq)]
+pub enum Operation {
+    // DeleteFile { path: String },
+    // DeleteDir,
+    // LinkFile,
+    Copy { from: String, to: String },
+    ReadFile { path: String },
+    CreateFile { path: String },
+    CreateDir { path: String },
+    OpenFile { path: String },
+    Write { path: String, content: String },
+}
+
+#[derive(Debug)]
+pub struct MockFilesystem {
+    copy_error:        Option<Box<dyn Error + Send + Sync>>,
+    create_dir_error:  Option<Box<dyn Error + Send + Sync>>,
+    create_file_error: Option<Box<dyn Error + Send + Sync>>,
+    open_file_error:   Option<Box<dyn Error + Send + Sync>>,
+    read_file_error:   Option<Box<dyn Error + Send + Sync>>,
+    write_error:       Option<Box<dyn Error + Send + Sync>>,
+    pub operations:    Vec<Operation>,
```

Comment:
Here you can extract a [type alias](https://doc.rust-lang.org/reference/items/type-aliases.html) to make the code cleaner, it can potentially be set in the ```shared.rs``` if reused across files:

```suggestion
type BoxError = Box<dyn Error + Send + Sync>; 

#[derive(Debug)]
pub struct MockFilesystem {
    copy_error:        Option<BoxError>
    create_dir_error:  Option<BoxError>
    create_file_error: Option<BoxError>
    open_file_error:   Option<BoxError>
    read_file_error:   Option<BoxError>
    write_error:       Option<BoxError>,
    pub operations:    Vec<Operation>,
```